File: 0394-decode-string/0394-decode-string.cpp
class Solution {
public:
    string decodeString(string &s, int &index){
        string dstring;
        int num = 0;
        
        while(index < s.length()){
            char ch = s[index];
            
            if(isdigit(ch)){
                num = num * 10 + (ch - '0');
            }else if(ch == '['){
                index++;
                string nestedstring = decodeString(s, index);
                for(int i = 0; i < num; i++){
                    dstring += nestedstring;
                }
                num = 0;
            }else if(ch == ']'){
                return dstring;
            }else{
                dstring += ch;
            }
            index++;
        }
        return dstring;
    }
    string decodeString(string s) {
        int index = 0;
        return decodeString(s, index);
    }
};

File: 0394-decode-string/0394-decode-string.py
class Solution:
    def decodeString(self, s: str) -> str:
        return self.decodeString1(s,0)

    def decodeString1(self, s, index):
        dstring = "" # correction here
        num = 0        
        while(index < len(s)):
            char = s[index]
            if char.isdigit():
                num = num*10 + int(char) # because int can be [1,300]
            elif char=='[':
                index+=1
                index, partstr = self.decodeString1(s,index)
                for i in range(num):
                    dstring+=partstr
                num=0
            elif char==']':
                return index,dstring
            else:
                dstring += char
            index+=1
        return dstring

        


        